### PR TITLE
Add lifetime outlives goal

### PR DESCRIPTION
**Pull request by jackh726**
c.c. #435 

I'm not sure if this is *exactly* the approach we want to take here, but I figured I would open this for comments at least

### CODE REVIEW

File: chalk-solve/src/clauses.rs [hunk 1]
```diff
@@ -244,6 +244,15 @@ fn program_clauses_that_could_match<I: Interner>(
                 .opaque_ty_data(opaque_ty.opaque_ty_id)
                 .to_program_clauses(builder),
         },
+        DomainGoal::Holds(WhereClause::LifetimeOutlives(a, b)) => {
+            builder.push_clause(
+                DomainGoal::Holds(WhereClause::LifetimeOutlives(a.clone(), b.clone())),
+                Some(DomainGoal::Holds(WhereClause::LifetimeOutlives(
```
Comment (nikomatsakis) [inline chalk-solve/src/clauses.rs:250]: This...doesn't seem right. This says

```
Holds('a: 'b) :- Holds('a: 'b)
```

which is kind of .. tautological.

**Comment by jackh726** [comment]
@nikomatsakis Okay so I rebased this to master. But I'm a little unsure how to continue here.

1) The `AddRegionConstraint` goal isn't really possible to do now that `chalk-engine` uses `Goal` directly
2) I looked at your branch in regards to having unification add goals instead of constraints. This doesn't quite fit into the current scheme since there's no translation there of `Outlives -> AddRegionConstraint`.
3) Can you elaborate on your last point? (applying `AddRegionConstraint` for things in the root universe, and higher-ranked things) Maybe a test case?

File: chalk-engine/src/simplify.rs [hunk 2]
```diff
@@ -68,14 +69,22 @@ impl<I: Interner, C: Context<I>> Forest<I, C> {
                     &goal.b,
                     &mut ex_clause,
                 )?,
-                GoalData::DomainGoal(domain_goal) => {
-                    ex_clause
-                        .subgoals
-                        .push(Literal::Positive(InEnvironment::new(
+                GoalData::DomainGoal(domain_goal) => match domain_goal {
```
Comment (nikomatsakis) [inline chalk-engine/src/simplify.rs:72]: it seems like this logic is missing from the recursive solver?

File: tests/test/misc.rs [hunk 3]
```diff
@@ -612,3 +612,33 @@ fn normalize_ambiguous() {
...
+            struct Bar {}
+
+            impl<'a, 'b> Foo<'a, 'b> for Bar where 'a: 'b {}
+        }
+
+        goal {
+            exists<'a, 'b> {
+                Bar: Foo<'a, 'b>
+            }
+        } yields[SolverChoice::slg(10, None)] {
```
Comment (nikomatsakis) [inline tests/test/misc.rs:630]: ah I guess that's why you have `slg` here

File: chalk-engine/src/simplify.rs (hunk 2 above)
Comment (nikomatsakis) [reply chalk-engine/src/simplify.rs:72]: I think you would add it here

https://github.com/rust-lang/chalk/blob/ffa2e515a4fac85f053dcbe6307d73ebc0e82043/chalk-solve/src/recursive/fulfill.rs#L174-L180

but it'd be nice to refactor the two solvers to share this logic.

**Comment by nikomatsakis** [comment]
@jackh726 

So, I think that your current branch is effectively "hard-coding" the `AddRegionConstraint` logic in some sense, by baking it into the "simplification" logic for handling domain goals (rather than adding a special sort of goal that is baked in, and then "converting" to that via a standard program clause). 

I don't quite understand your point about having refactored the `resolvent` special handling, and I don't quite understand what you mean about not being able to modify unification, but I can maybe try to do the thing I was talking about and see what happens.

I *think* I still prefer to have a special goal like `AddRegionConstraint` (or maybe a built-in notion of `Outlives`) so that the solver never tries to "bake in" what a domain-goal means, but it's not a really big difference.

As for the interaction with higher-ranked, let's shelve it for now, I don't think it's relevant to this PR yet.

**Pull request merged**